### PR TITLE
Company Ai Functions: ThinkState3

### PR DESCRIPTION
**Pull request by duncanspumpkin**
As discussed in the team meeting ai is not the priority so won't take this any further for now. It looks like think state 3 is placing track and then signals and finally considers creating a station. I've only done some of the more simple functions at the start but there are some easy ones in there that could also be done for the signals in the future. I had to rename Math::Vector::distance to distanceXY as we have some users that take a Pos3 and only want the XY distance and the ai functions wants the XYZ distance. Other names could be distance2D and distance3D.

### CODE REVIEW

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 1]
```diff
@@ -323,12 +325,307 @@ namespace OpenLoco
...
+        company.var_85F0 = 0;
+        company.var_85EE = 0;
+        company.var_85EF = 0;
+        company.var_85DE = 0;
+        company.var_85E2 = 0;
+        company.var_85E8 = 0;
+
+        const auto distance = std::min<uint16_t>(256, Math::Vector::distanceXYZ(World::Pos3(company.var_85C4, company.var_85C8 * World::kSmallZStep), World::Pos3(company.var_85C9, company.var_85CD * World::kSmallZStep)));
+        company.var_85EA = distance / 2 + distance * 2;
+        std::fill(std::begin(company.var_25C0), std::end(company.var_25C0), 0xFFFFU);
```
Comment (duncanspumpkin) [inline src/OpenLoco/src/World/CompanyAi.cpp:504]: this probably will fail to build...

**Comment by AaronVanGeffen** [comment]
I'm fine with this as-is, but just to be on the safe side: did you run a replay test involving AI to ensure there's no regressions?

**Comment by duncanspumpkin** [comment]
End company has now been done under #2283.

**Review by AaronVanGeffen — APPROVED** [review]
Happy to expand on our knowledge of the inner workings of the Locomotion AI. If it doesn't diverge, I'm fine with this being merged as-is.

**Pull request merged**